src/main/java/org/myrobotlab/codec/CodecUtils.java

```diff
@@ -1244,61 +1247,75 @@ public static void main(String[] args) {
-         */
-        public final String path; // {scheme}://{host}:{port}/api/messages
+    /**
+     * The path to reach this API
+     */
+    public final String path; // {scheme}://{host}:{port}/api/messages
 
-        /**
-         * An example URI to reach this API
-         */
-        public final String exampleUri;
+    /**
+     * An example URI to reach this API
+     */
+    public final String exampleUri;
 
-        /**
-         * The description of this API
-         */
-        public final String description;
+    /**
+     * The description of this API
+     */
+    public final String description;
 
-        /**
-         * Construct a new API description.
-         *
-         * @param key            {@link #key}
-         * @param uriDescription {@link #path}
-         * @param exampleUri     {@link #exampleUri}
-         * @param description    {@link #description}
-         */
-        public ApiDescription(String key, String uriDescription, String exampleUri, String description) {
-            this.key = key;
-            this.path = uriDescription;
-            this.exampleUri = exampleUri;
-            this.description = description;
-        }
+    /**
+     * Construct a new API description.
+     *
+     * @param key
+     *          {@link #key}
+     * @param uriDescription
+     *          {@link #path}
+     * @param exampleUri
+     *          {@link #exampleUri}
+     * @param description
+     *          {@link #description}
+     */
+    public ApiDescription(String key, String uriDescription, String exampleUri, String description) {
+      this.key = key;
+      this.path = uriDescription;
+      this.exampleUri = exampleUri;
+      this.description = description;
+    }
 
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            ApiDescription that = (ApiDescription) o;
-            return Objects.equals(key, that.key)
-                    && Objects.equals(path, that.path)
-                    && Objects.equals(exampleUri, that.exampleUri)
-                    && Objects.equals(description, that.description);
-        }
+    @Override
+    public boolean equals(Object o) {
+      if (this == o)
+        return true;
+      if (o == null || getClass() != o.getClass())
+        return false;
+      ApiDescription that = (ApiDescription) o;
+      return Objects.equals(key, that.key) && Objects.equals(path, that.path) && Objects.equals(exampleUri, that.exampleUri) && Objects.equals(description, that.description);
+    }
 
-        @Override
-        public int hashCode() {
-            return Objects.hash(key, path, exampleUri, description);
-        }
+    @Override
+    public int hashCode() {
+      return Objects.hash(key, path, exampleUri, description);
     }
+  }
+
+  /**
+   * Single parameter from JSON. Will use default return type, currently LinkedHashMap

```

Comment:
I'm not sure I'm following you AP.  
And its possible I munged my intended implementation of updates ... so lets start with "intent"
My intent is: 
Jackson, GSON, or whatever library which provides the parsing details should from the "outside" behave the same.
This is part of the intent of CodeUtils.

The act of parsing - we either do, or do not get a suggested type.  
If we do not, then we need a default return type from the parsing.
The appropriate return type offered to the publish must be a native java container and it must be devoid of any libraries implementation.   If there is downstream cod the utilizes CodecUtils.fromJson but requires it to be library specific, then 
1. it should be private
2. it needs to implement something besides fromJson for its own.  fromJson "is" public

Reply:
I assume this comment is in reference to the above change to GSON_DEFAULT_OBJECT_TYPE as well as the linked javadoc. I agree that we want CodecUtils to provide the same interface and we want a default return type. The issue is that the GSON_DEFAULT_OBJECT_TYPE is *not* for that purpose, it is used privately to determine what the current JSON_DEFAULT_OBJECT_TYPE is, which is simply an indicator of what our selected deserializer will deserialize to when you pass in `Object`, irrespective of CodecUtils. This is used for checking whether a field was double encoded, as used in the method cache here:

https://github.com/MyRobotLab/myrobotlab/blob/e035de0a6345a4e4a40d16076b13c671f936248a/src/main/java/org/myrobotlab/framework/MethodCache.java#L554

Thus, by changing how GSON_DEFAULT_OBJECT_TYPE is defined, we will always skip over that branch and cause errors. We do not have control over this type because GSON skips our custom deserializer for Object declarations, and since the field is embedded inside a Message we cannot intercept the requested return type in CodecUtils. It is hard coded inside GSON itself. If we want to remove dependence on that and figure out a different way to determine whether parameters were double encoded or not I'm perfectly okay with fixing it, but that might need to be done in a different PR.

For the linked doccomment, the issue is purely from the documentation side. Usually we would want to steer clear of documenting a single concrete implementation as the return type, because that may result in code downstream (or in external projects, however unlikely that may be) depending on that choice because it's in the documentation and thus it is assumed to be a part of the method contract. If we instead put in the documentation that the method returns a `Map`, most developers would only depend on the features provided by the map interface and we would be free to change implementations later, for example if we wanted to use a `ConcurrentHashMap` instead. Additionally, the code that the doccomment is documenting does not actually explicitly deserialize to the map, instead it relies on a side effect of a secondary method. If that side effect is changed, this method will also change and the documentation most likely not updated to reflect the change. If we want a concrete implementation as the default, then that default should be placed in a constant and used instead of the null value (and instead of the inline `LinkedHashMap.class` in the other overload). If you want that type to be constant no matter the deserializer implementation, then it *needs* to be a new constant and not `GSON_DEFAULT_OBJECT_TYPE` or `JSON_DEFAULT_OBJECT_TYPE`

Reply:
In short, you want the deserializer to take advantage of the meta data that is currently being put into the json ?
And my current assumption of type when not supplied defeats that ?

Reply:
No, the problem doesn't have anything to do with the class metadata, the code that is breaking due to this change was present long ago but referenced `LinkedTreeMap` directly. I just pulled that reference into a static and called it `GSON_DEFAULT_OBJECT_TYPE`. Here's an example that will hopefully help:
```java
// See that json's data is intentionally not double encoded
String json = "{\"name\":\"test\", \"method\":\"test_method\", \"data\":[{\"param1\":1}]}";

// First-phase decode
Message msg = CodecUtils.fromJson(json, Message.class);

// Now, if we decided to just run through the second-phase decode, it would blow up
// because msg's data is not an array of Strings
// The current code checks this condition by checking the type of the data first
// Since we cannot intercept the deserialization of the data (limitation of GSON),
// we will always get a data of LinkedTreeMap for objects, regardless of anything else we do.
// The code thus checks the data against the LinkedTreeMap, and since it cannot be changed
// we know if we need to do second phase decoding differently (the current code re-serializes the map
// before doing the second phase decode)
for (int i = 0; i < msg.data.length; i++) {
    if (LinkedTreeMap.class.isAssignableFrom(msg.data[i].getClass()) {
        msg.data[i] = CodecUtils.toJson(msg.data[i]);
    }

   // Continue deser
}

```

Another way to do this check is to invert it, and re-serialize everything that is *not* a String. This would probably be safer because the current implementation does not work for other primitives and would allow you to use the constants how you see fit. But that requires reworking the current logic to do so.

Reply:
So I bork'd things by accidently changing default to LinkedHashMap ? 
I should change it back to LinkedTreeMap ?

Reply:
Yeah, and then for your defaults, i.e. what type we use when null is passed into `CodecUtils.fromJson(()`, can be made another constant, like `DEFAULT_OBJECT_TYPE` or something like that

Reply:
How's this ? - https://github.com/MyRobotLab/myrobotlab/commit/b6ec557719f10879be87ce2a31ae0583e291a7f4

Reply:
Close, `JACKSON_DEFAULT_OBJECT_TYPE` was actually fine because `LinkedHashMap` is what Jackson uses as its default while GSON uses `LinkedTreeMap`. Line 283 should instead link to `DEFAULT_OBJECT_TYPE` so the documentation stays updated, and line 295 should replace `LinkedHashMap.class` with `DEFAULT_OBJECT_TYPE`. Finally, since `DEFAULT_OBJECT_TYPE` is now decoupled from the backends, you can freely change it to `LinkedHashMap`

Reply:
there is no references at all to LinkedHashMap in CodecUtils .. but I've updated the javadoc link now

Reply:
There were references, https://github.com/MyRobotLab/myrobotlab/commit/b6ec557719f10879be87ce2a31ae0583e291a7f4 removed them

Reply:
all good on the head ?

Reply:
#1253 changes the LinkedTreeMaps to LinkedHashMap where appropriate, after that this issue should be good

Reply:
GSON is dead at this point isn't it ?  I mean it should be removed .. that would be the plan going forward.. right ?

Reply:
It will be removed, eventually, but I'd really like to keep it as an emergency backup if Jackson isn't quite configured right yet. So far, things are looking good and Jackson seems to be configured correctly, but if we run into an issue with it in a month or something like that, then we need to figure out the issue and push a hotfix as quickly as possible instead of just switching back to GSON and having time to fix it

Reply:
merged - branch ap-small-codec-fix